hack/build.sh

```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
+
+# Script to compile the agent binary for different architectures. To add a new architecture, create a new function named
+# compile_<GOOS>_<GOARCH. Within that function, set the GOOS and GOARCH variables for the architecture you want, then
+# call the compile function. Binaries are output in ./builds/<GOOS>/<GOARCH>/
+
+BIN_NAME=preflight
+
+# Build information that is compiled into the binaries.
+COMMIT=$(git rev-list -1 HEAD)
+DATE="$(date -uR)"
+VERSION="$(bash ./hack/version.sh)"
+
+# OAuth2 config for the agent to work with platform.jetstack.io
+OAUTH_CLIENT_ID=k3TrDbfLhCgnpAbOiiT2kIE1AbovKzjo
+OAUTH_CLIENT_SECRET=f39w_3KT9Vp0VhzcPzvh-uVbudzqCFmHER3Huj0dvHgJwVrjxsoOQPIw_1SDiCfa
+OAUTH_AUTH_SERVER_DOMAIN=auth.jetstack.io
+
+GOOS=""
+GOARCH=""
+GOARM=""
+
+function compile_all() {
+  compile_linux_amd64
+  compile_linux_arm64
+  compile_linux_armv7
+  compile_darwin_amd64
+  compile_windows_amd64
+}
+
+function compile_windows_amd64() {
+  GOOS=windows
+  GOARCH=amd64
+
+  compile
+}
+
+function compile_darwin_amd64() {
+  GOOS=darwin
+  GOARCH=amd64
+
+  compile
+}
+
+function compile_linux_armv7() {
+  GOOS=linux
+  GOARCH=arm
+  GOARM=7
+  compile
+  GOARM=""
+}
+
+function compile_linux_arm64() {
+  GOOS=linux
+  GOARCH=arm64
+
+  compile
+}
+
+function compile_linux_amd64() {
+  GOOS=linux
+  GOARCH=amd64
+
+  compile
+}
+
+function compile() {
+  OUTPUT=./builds/${GOOS}/${GOARCH}/${BIN_NAME}
+
+  # If GOARM has been set, add an extra directory.
+  if [ -n "$GOARM" ]; then
+    OUTPUT=./builds/${GOOS}/${GOARCH}/v${GOARM}/${BIN_NAME}
+  fi
+
+  GOOS=${GOOS} GOARCH=${GOARCH} GOARM=${GOARM} CGO_ENABLED=0 go build -ldflags \
+  "-w -s "\
+"-X 'github.com/jetstack/preflight/pkg/version.PreflightVersion=${VERSION}'"\
+"-X 'github.com/jetstack/preflight/pkg/version.Platform=${GOOS}/${GOARCH}'"\
+"-X 'github.com/jetstack/preflight/pkg/version.Commit=${COMMIT}'"\
+"-X 'github.com/jetstack/preflight/pkg/version.BuildDate=${DATE}'"\
+"-X 'github.com/jetstack/preflight/pkg/version.GoVersion=${GOVERSION}'"\
+"-X 'github.com/jetstack/preflight/pkg/client.ClientID=${OAUTH_CLIENT_ID}'"\
+"-X 'github.com/jetstack/preflight/pkg/client.ClientSecret=${OAUTH_CLIENT_SECRET}'"\
+"-X 'github.com/jetstack/preflight/pkg/client.AuthServerDomain=${OAUTH_AUTH_SERVER_DOMAIN}'" \
+  -o "${OUTPUT}"
+}
+
+compile_all
```

Comment:
Personally I prefer having this and `docker.sh` in the Makefile itself:

```
OAUTH_CLIENT_ID?=k3TrDbfLhCgnpAbOiiT2kIE1AbovKzjo
OAUTH_CLIENT_SECRET?=f39w_3KT9Vp0VhzcPzvh-uVbudzqCFmHER3Huj0dvHgJwVrjxsoOQPIw_1SDiCfa
OAUTH_AUTH_SERVER_DOMAIN?=auth.jetstack.io

define LDFLAGS
-X "github.com/jetstack/preflight/pkg/version.PreflightVersion=$(VERSION)" \
-X "github.com/jetstack/preflight/pkg/version.Platform=$(GOOS)/$(GOARCH)" \
-X "github.com/jetstack/preflight/pkg/version.Commit=$(COMMIT)" \
-X "github.com/jetstack/preflight/pkg/version.BuildDate=$(DATE)" \
-X "github.com/jetstack/preflight/pkg/version.GoVersion=$(GOVERSION)" \
-X "github.com/jetstack/preflight/pkg/client.ClientID=$(OAUTH_CLIENT_ID)" \
-X "github.com/jetstack/preflight/pkg/client.ClientSecret=$(OAUTH_CLIENT_SECRET)" \
-X "github.com/jetstack/preflight/pkg/client.AuthServerDomain=$(OAUTH_AUTH_SERVER_DOMAIN)"
endef

GO_BUILD:=CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)'
GO_INSTALL:=CGO_ENABLED=0 go install -ldflags '$(LDFLAGS

.PHONY: ./builds/$(GOOS)/$(GOARCH)/$(BIN_NAME)
./builds/$(GOOS)/$(GOARCH)/$(BIN_NAME):
	GOOS=$(GOOS) GOARCH=$(GOARCH) $(GO_BUILD) -o ./builds/$(GOOS)/$(GOARCH)/$(BIN_NAME) .
.PHONY: ./builds/$(GOOS)/$(GOARCH)/v$(GOARM)/$(BIN_NAME)
./builds/$(GOOS)/$(GOARCH)/v$(GOARM)/$(BIN_NAME):
	GOOS=$(GOOS) GOARCH=$(GOARCH) GOARM=$(GOARM) $(GO_BUILD) -o ./builds/$(GOOS)/$(GOARCH)/v$(GOARM)/$(BIN_NAME) .

build-all-platforms:
	$(MAKE) GOOS=linux   GOARCH=amd64       ./builds/linux/amd64/$(BIN_NAME)
	$(MAKE) GOOS=linux   GOARCH=arm64       ./builds/linux/arm64/$(BIN_NAME)
	$(MAKE) GOOS=linux   GOARCH=arm GOARM=7 ./builds/linux/arm/v7/$(BIN_NAME)
	$(MAKE) GOOS=darwin  GOARCH=amd64       ./builds/darwin/amd64/$(BIN_NAME)
	$(MAKE) GOOS=windows GOARCH=amd64       ./builds/windows/amd64/$(BIN_NAME)

# Docker image
PLATFORMS?=linux/arm/v7,linux/arm64/v8,linux/amd64
BUILDX_EXTRA_ARGS?=

push_buildx_args=--push $(BUILDX_EXTRA_ARGS)
push-canary_buildx_args=--tag $(DOCKER_IMAGE):canary --push $(BUILDX_EXTRA_ARGS)
build_buildx_args=$(BUILDX_EXTRA_ARGS)

.PHONY: _docker-%
_docker-%: build-all-platforms
	docker buildx build --platform $(PLATFORMS) \
	--tag $(DOCKER_IMAGE_TAG) \
	$($*_buildx_args) \
	.

build-docker-image: _docker-build
push-docker-image: _docker-push

```

I find it quicker to read and understand and easier to use:

- With `make`, if I run `make build-docker-image`, it knows it needs to build the binaries as well (as the dependency is encoded in the Makefile). You could mitigate that in your approach by adding a dep in the Makefile as well.

- Ideally, we want to reduce the differences between what CI does and what we can do locally, so relying on something like `make` to orchestrate the different steps (binary build + docker image) is better than having to execute several commands in certain order.

- Then, you someone is testing this repo locally, the common actions are going to be: either build a binary for their platform to try their change or build a docker image to try changes in things like Dockerfile (also in one platform first, so you can iterate faster). With our previous approach with make they can do `make build` and build just a binary or `make build-docker-image PLATFORM=linux/amd64`  and build just one docker image (and not push it).

- BONUS: with make you have input args autocompletion -> `make build-docker-image PLATF<tab>`.

Reply:
I find the Makefile better for those reasons, and it is not hard to read to me, but I am biased because I use Makefiles a lot.

If you and others think using scripts is simples and easier to maintain and understand, I have no problems using scripts instead.

@tfadeyi @charlieegan3 WDYT?

Reply:
Fair, I think part of the problem I am having with it is I am not that familiar with how Makefiles work outside of just creating targets and making them call other targets beforehand. I can't look at the Makefile as is and instantly understand how the docker builds work.

Especially with sections like this:

```
.PHONY: ./builds/$(GOOS)/$(GOARCH)/$(BIN_NAME)
./builds/$(GOOS)/$(GOARCH)/$(BIN_NAME):
	GOOS=$(GOOS) GOARCH=$(GOARCH) $(GO_BUILD) -o ./builds/$(GOOS)/$(GOARCH)/$(BIN_NAME) .
.PHONY: ./builds/$(GOOS)/$(GOARCH)/v$(GOARM)/$(BIN_NAME)
./builds/$(GOOS)/$(GOARCH)/v$(GOARM)/$(BIN_NAME):
	GOOS=$(GOOS) GOARCH=$(GOARCH) GOARM=$(GOARM) $(GO_BUILD) -o ./builds/$(GOOS)/$(GOARCH)/v$(GOARM)/$(BIN_NAME) .
```

To me this is just incredibly cryptic, and doesn't feel necessary.

Reply:
Let's inspect that section in particular block by block:

```
.PHONY: ./builds/$(GOOS)/$(GOARCH)/$(BIN_NAME)
```
This just means that a file like `./builds/linux/arm64/preflight` is going to be recreated everytime, even if it already exists. I would be up for removing this one in particular but I thought it might avoid confusions.

```
./builds/$(GOOS)/$(GOARCH)/$(BIN_NAME):
    GOOS=$(GOOS) GOARCH=$(GOARCH) $(GO_BUILD) -o ./builds/$(GOOS)/$(GOARCH)/$(BIN_NAME) .
```

This one defines how to build `./builds/linux/arm64/preflight` -> `GOOS=linux GOARCH=arm64 go build ... -o ./builds/linux/arm64/preflight`


```
.PHONY: ./builds/$(GOOS)/$(GOARCH)/v$(GOARM)/$(BIN_NAME) ./builds/$(GOOS)/$(GOARCH)/v$(GOARM)/$(BIN_NAME): GOOS=$(GOOS) GOARCH=$(GOARCH) GOARM=$(GOARM) $(GO_BUILD) -o ./builds/$(GOOS)/$(GOARCH)/v$(GOARM)/$(BIN_NAME) .
```

This last block is the same as the previous ones but matching the case of builds with `GOARM`. There must be ways of merging the two blocks but maybe that goes too much to the black magic side of things.

Reply:
I should have added more documentation in the Makefile in the first place.

Reply:
Thanks for the explanation 😄. My issue is that just to understand what these 5/6 lines do requires explanation is part of the problem. For anyone unfamiliar to make I think it will just come off as intimidating and hard to parse. 

By delegating to scripts, we can use concepts we're familiar with already, like functions and control flow. Perhaps I just need to level up my `Makefile` abilities, but from a perspective of developer-friendliness I'm just not sure this is the best way.

Reply:
I think the script is pretty clear, but I do like being able to build only for amd locally, could we add a way of just building the binary for amd as a default when calling the script?

Reply:
I confess that I also find the script easier to read. However, I'd like to just check what we need.

I actually build local binaries with go build manually. I don't use the makefile. Generally I use go run locally. If I needed a binary, I could work out how to make one the same from a makefile or script.

I'd like to find a system that allows us to use the same binaries in the image as we push to the release. Or have the release binaries build separately from pushing the image in a concurrent action.

I'd also like to see this split out from the actual fix: https://github.com/jetstack/preflight/pull/230/files#r621265094